### ISSUE TITLE
"CustomError enum doesn't Display" false positive error using "failure" create

### ISSUE THREAD

**Issue by lancelote** [intellij-rust/intellij-rust#5954]
## Environment

* **IntelliJ Rust plugin version:** 0.3.130.3293-202-nightly
* **Rust toolchain version:** 1.47.0-nightly (6c8927b0c 2020-07-26) x86_64-apple-darwin
* **IDE name and version:** CLion 2020.3 Snapshot (CL-203.2576)
* **Operating system:** macOS 10.15.6

## Problem description

With [failure](https://crates.io/crates/failure) crate display is defined in the following way

```rust
#[derive(Debug, Fail)]
pub enum DocumentServiceError {
    #[fail(display = "You have exceeded the allowed number of documents per minute.")]
    RateLimitExceeded(Backtrace),
    #[fail(display = "I/O error: {}", _0)]
    IO(io::Error, Backtrace),
}
```

Unfortunately, the plugin doesn't understand it and reports false positive error

<img width="955" alt="error" src="https://user-images.githubusercontent.com/2500320/90500193-3969ed80-e153-11ea-925e-1971a08513d3.png">

## Steps to reproduce

```rust
extern crate failure;
extern crate failure_derive;

use failure::{Backtrace, Fail};
use std::fs::File;
use std::fs::OpenOptions;
use std::io;
use std::result;

const MAX_DOCS_CREATED_PER_MINUTE: u8 = 100;

fn num_documents_created_in_last_minute() -> u8 {
    2
}

#[derive(Debug, Fail)]
pub enum DocumentServiceError {
    #[fail(display = "You have exceeded the allowed number of documents per minute.")]
    RateLimitExceeded(Backtrace),
    #[fail(display = "I/O error: {}", _0)]
    IO(io::Error, Backtrace),
}

impl From<io::Error> for DocumentServiceError {
    fn from(other: io::Error) -> Self {
        DocumentServiceError::IO(other, Backtrace::new())
    }
}

// Simplify return type wording
pub type Result<T> = result::Result<T, DocumentServiceError>;

pub fn create_document(filename: &str) -> Result<File> {
    if num_documents_created_in_last_minute() > MAX_DOCS_CREATED_PER_MINUTE {
        return Err(DocumentServiceError::RateLimitExceeded(Backtrace::new()));
    }

    let file = OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(filename)?;

    Ok(file)
}

fn create_project(project_name: &str) -> Result<()> {
    create_document(&format!("{}-draft1", project_name))?;
    create_document(&format!("{}-draft2", project_name))?;
    create_document(&format!("{}-revision1", project_name))?;
    create_document(&format!("{}-revision2", project_name))?;
    Ok(())
}

fn main() {
    match create_project("my-project") {
        Ok(()) => println!("Project created successfully"),
        Err(e) => {
            println!("Project creation failed: {}", e);
            if let Some(backtrace) = e.backtrace() {
                if !backtrace.to_string().trim().is_empty() {
                    println!("{:?}", backtrace);
                }
            }
        }
    }
}
```

... with dependencies

```
failure = "0.1.8"
failure_derive = "0.1.8"
backtrace = "0.3.50"
```

**Comment by LoganDark** [comment]
Quick note, the `failure` crate is a very old library, designed to be as incompatible as possible with error types from any other crate. The `anyhow` crate is a newer library that actually works *with* existing Rust libraries rather than against them.

**Comment by qb60** [comment]
> Quick note, the `failure` crate is a very old library, designed to be as incompatible as possible with error types from any other crate. The `anyhow` crate is a newer library that actually works _with_ existing Rust libraries rather than against them.
> 
> If you just need to define error types, though, and not a special type that can catch any error (which `anyhow::Error` is), `thiserror` is probably what you want.

The same with thiserror for me:
`use thiserror::Error;

#[derive(Error, Debug)]
enum Error {
    #[error("Error 1")]
    SomeError,
}

fn foo() -> Result<(), Error> {
    Err(Error::SomeError)
}

fn main() {
    if let Err(e) = foo() {
        eprintln!("Error {} occured", e)
    }
}`

`[dependencies]
thiserror = "1.0.20"`

![Выделение_007](https://user-images.githubusercontent.com/1970596/91526508-3920dd80-e91d-11ea-913b-f464f6d2a30d.png)

**Comment by LoganDark** [comment]
@qb60 That is definitely a bug with intellij-rust because thiserror certainly does implement Display. Try to find an existing issue or open one maybe?

**Comment by afidegnum** [comment]
i'm also facing a similar issue.

**Comment by jjmark15** [comment]
I also experience this issue with `thiserror`. No compilation issues, only Intellij Rust false positive error. Image shows warning for an error that is implemented using `thiserror`.

![Screenshot (7)](https://user-images.githubusercontent.com/11464367/96505946-5c457a80-1246-11eb-84f8-3e1b0bcc1960.png)

**Comment by afidegnum** [comment]
is there any way around ?

**Comment by ronlobo** [comment]
Is this related?

<img width="436" alt="image" src="https://user-images.githubusercontent.com/379246/98873879-1cc02600-24b4-11eb-96d1-3d0ad4a9af3e.png">

**Comment by dimlev** [comment]
I also experience this issue with thiserror. 

It is very annoying since i am getting false error indicators all over the place and makes it hard to find actual errors.

Is there any way around this? Perhaps fake it or ignore it?

**Comment by PiDelport** [comment]
Related issue:
- #5478

**Comment by PiDelport** [comment]
For folks looking for a workaround, enabling the `org.rust.cargo.evaluate.build.scripts` and `org.rust.macros.proc` experimental features as described in #6908 seems to fix the issue. 🎉️
